app/js/arethusa.core/services/state.js

```diff
@@ -452,6 +452,36 @@ angular.module('arethusa.core').service('state', [
       selectSurroundingToken('prev');
     };
 
+    this.getPreviousTokens = function(id,numTokens) {
+      var tokens = [];
+      var allIds = Object.keys(self.tokens);
+      var endIndex = allIds.indexOf(id) - 1;
+      if (endIndex >= 0) {
+        var startIndex = 0;
+        if (numTokens && startIndex !== endIndex) {
+          startIndex = startIndex + numTokens;
+        }
+        for (var i=startIndex; i<= endIndex; i++) {
+          tokens.push(self.getToken(allIds[i]));
+        }
+      }
+      return tokens;
+    };
+
+    this.getNextTokens = function(id,numTokens) {
```

Comment:
ah, thanks these are good catches. This is sloppy refactoring on my part. I originally coded it so that it always returned the requested number of tokens, and then realized I needed something that either returned just the previous 1 token, or ALL of the previous tokens in the sentence, so I made the numTokens parameter optional but didn't do it quite right. I will fix this. Thank you!

Reply:
ok, will look at fixing this too. I thought I had tested it well enough but I guess not :-(